Fix attach goroutine/fd leak when no I/O is ready

cpuguy83 opened the pull request:
When only stdin is attached the goroutine can only ever exit if:

1. The container pipe is closed while trying to write to it
2. The client closes the stdin read pipe

This is because `io.Copy` does a read on the read side then a write to the write side.
If reading from the client's stdin pipe blocks, the goroutine will never get notified that the container pipe is closed.

Another similar case is when the client disconnects but `io.Copy` is blocking waiting for data from the container's stdout/stderr streams.
This is significantly trickier to resolve.
The 2nd commit fixes this for Linux only.

- Fixes #37182

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
fix a goroutine and file-descriptor leak on container attach
```

**- A picture of a cute animal (not mandatory but encouraged)**

cpuguy83 committed:
Fix fd leak/goroutine when attaching stdin only

When only stdin is attached the goroutine can only ever exit if:

1. The container pipe is closed while trying to write to it
2. The client closes the stdin read pipe

This is because `io.Copy` does a read on the read side then a write to
the write side.
If reading from the client's stdin pipe blocks, the goroutine will never
get notified that the container pipe is closed.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Fix goroutine/fd leak when client disconnects

In cases where the client disconnects and there is nothing to read from
a stdio stream after that disconnect, the copy goroutines and file
descriptors are leaked because `io.Copy` is just blocked waiting for
data from the container's I/O stream.

This fix only applies to Linux.
Windows will need a separate fix.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented on api/server/router/container/container_routes.go:
```diff
@@ -736,6 +738,8 @@ func (s *containerRouter) wsContainersAttach(ctx context.Context, w http.Respons
 			conn.PayloadType = websocket.BinaryFrame
 		}
+
+		// TODO: Close notifications
```
In order to do this we need to switch to a lower-level websocket library such as https://pkg.go.dev/github.com/gobwas/ws

thaJeztah commented:
/cc @corhere

corhere reviewed (commented):
I wonder if the client-disconnects solution is more clever than it needs to be. In the case of TCP, the lowest common denominator for connection state tracking, each peer can either gracefully close its write stream (`FIN`) or forcefully abort the whole connection (`RST`) but there is no affordance in the TCP state machine to signal gracefully that a peer is no longer interested in _receiving_ any more data. And the only portable way to check whether a socket is still open is to try reading or writing at least one byte and see if it succeeds. What if we always used the client-closed-for-writing signal as our cue to stop copying stdout/stderr from the container? That is, have a goroutine which calls `conn.Read()` even when stdin isn't attached to the container, and closes stdout/stderr when the read returns `io.EOF` or similar. The same code path would work for UNIX domain sockets, TCP sockets, WebSockets and Windows named pipes; no platform-specific event handling required! And all the bases would be covered for TCP: FIN, RST and keep-alive timeout will all result in EOF on read.

cpuguy83 commented:
@corhere You are right I think this could work.
I've added a 3rd commit with the changes.

thaJeztah commented on container/stream/attach_test.go:
```diff
@@ -0,0 +1,102 @@
+	"gotest.tools/v3/assert"
+)
+
+// Make sure when there is no I/O on a stream that the goroutines do not get blcoked after the container exits.
```
🙈 

```suggestion
// Make sure when there is no I/O on a stream that the goroutines do not get blocked after the container exits.
```

Wondering if there's more to describe for this test; would it make sense to add some links to the GitHub ticket(s)? (mostly to find details in future if needed)

thaJeztah replied:
@cpuguy83 don't forget this one if you need to push again 😅

cpuguy83 replied:
Fixed

corhere commented:
Sigh, [the CLI half-closes the connection to signal EOF on stdin.](https://github.com/docker/cli/blob/de0d30ff24fad55df89adcb3c01d10bb1821c8db/cli/command/container/hijack.go#L171) So much for that idea. It's looking like the sorts of OS-specific close notifications you had implemented initially is really the only thing we could do to fix the leaks within the current API versions.

Using socket half-close to signal stdin closed rather than attach-done is arguably a design flaw in our container-attach protocol. With the WebSocket attach protocol also busted (ws provides framing but no multiplexing so stdout can't be distinguished from stderr) I see an opportunity to fix all these flaws... the third time around. I'm imagining new WebSocket attach protocol (paired with new endpoints?) for attach and exec which:
- multiplexes output streams in a way which takes advantage of WebSocket framing (i.e. not `stdcopy`)
- unconditionally multiplexes the output streams even when there is only a single output stream so that the daemon could signal the client out-of-band
- also multiplexes stdin with out-of-band signaling for stdin-EOF and perhaps other things (`SIGWINCH` without making a separate request?)
But I digress...

cpuguy83 commented:
Rebased this and removed the 3rd commit which didn't work as hoped.

corhere commented on api/server/router/container/notify_linux.go:
```diff
@@ -0,0 +1,53 @@
+		}
+
+		events := make([]unix.EpollEvent, 1)
+		if _, err := unix.EpollWait(epFd, events, -1); err != nil {
+			log.G(ctx).WithError(err).Warn("notifyClosed: failed to wait for close notifications")
+			return
+		}
```
Doesn't this need to be called in an EINTR loop to gracefully handle getting interrupted by a signal?
```suggestion
		for {
			_, err := unix.EpollWait(epFd, events, -1)
			if err != nil {
				if errors.Is(err, unix.EINTR) {
					continue
				}
				log.G(ctx).WithError(err).Warn("notifyClosed: failed to wait for close notifications")
				return
			}
			break
		}
```

cpuguy83 replied:
Fixed

corhere reviewed (commented):
I'm not super thrilled about the un-idiomatic use of callbacks in `notifyClosed` but I can understand why: the more idiomatic designs (synchronous blocking call returning an error, return a child context in the same vein as ["os/signal".NotifyContext](https://pkg.go.dev/os/signal#NotifyContext)) would take a nontrivial amount of additional plumbing and refactoring. I have to concede that you have gone with the most pragmatic implementation.

LGTM!

cpuguy83 commented on api/server/router/container/notify_unsupported.go:
```diff
@@ -0,0 +1,10 @@
+	"net"
+)
+
+func notifyClosed(conn net.Conn, notify func()) {}
```
Fixed

corhere reviewed (approved):
:shipit:

cpuguy83 commented:
Rebased and moved EINTR handlers to the unix_noeintr package.

neersighted commented on daemon/attach.go:
```diff
@@ -52,10 +52,27 @@ func (daemon *Daemon) ContainerAttach(prefixOrName string, c *backend.ContainerA
+		}
+	}()
+
+	inStream, outStream, errStream, err := c.GetStreams(multiplexed, closeNotify)
```
Can we get a FIXME here that notes we should properly plumb contexts instead of passing a notification callback?

cpuguy83 replied:
I'm not sure what you mean here.
At best we would need to pass a context back rather than a passing in `closeNotify`.

e.g. `ctx, inStream, outStream, errStream, err := c.GetStreams(ctx, multiplexed)`

`c` here is not the container, its the attach config (and I suppose poorly referenced).

cpuguy83 replied:
Not sure if this is the actual confusing here, but I went ahead and renamed `c` to `req` because `c` is usually `container`, and there is also `ctr` and `cfg`... so lots of things to be confused on when reading the code in general.

neersighted commented on api/server/router/container/notify_unsupported.go:
```diff
@@ -0,0 +1,11 @@
+//go:build !linux
+// +build !linux
```
No need to use the legacy buildtags anymore.

cpuguy83 replied:
Fixed.

neersighted reviewed (approved):
LGTM (with a couple nits), with a bit of being sad about not being able to solve this on Windows, but I don't begrudge not wanting to figure out a separate async I/O API.

cpuguy83 committed:
refactor: rename attach config var

This var for the incoming attach request.
Just within this one function we also have `cfg`, and `ctr` already, so
`c` just makes things more confusing.
Not to mention `c` is usually referencing a container object in other
parts of the code.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented:
Failure on rootless looks related; https://github.com/moby/moby/actions/runs/8428203898/job/23080313424?pr=45052#step:8:1254

```
=== FAIL: amd64.integration.container TestAttachDisconnectLeak (0.07s)
    attach_test.go:102: assertion failed: infoAttach.NGoroutines is <= info.NGoroutines
```


I kicked CI again to see if it's flaky

cpuguy83 commented:
Updated the integration test to hopefully make it less likely to flake.

cpuguy83 commented:
Ok I think the test issue was that it was using the main integration daemon and the goroutine count was fluctuating at just the right times to break it.
I've switched it to use to use its own daemon to help reduce noise and that seems to have done the trick.

cpuguy83 commented:
Ping, does this LGTY?

cpuguy83 commented:
btw, I've opened a proposal on Go to extend the syscall.RawConn interface (or add an additional interface) to receive close notifications.
https://github.com/golang/go/issues/67432

cpuguy83 commented:
Ping

vvoland commented on api/server/router/container/notify_linux.go:
```diff
@@ -0,0 +1,54 @@
+	defer unix.Close(epFd)
+
+	err = rc.Control(func(fd uintptr) {
+		err := unix_noeintr.EpollCtl(epFd, unix.EPOLL_CTL_ADD, int(fd), &unix.EpollEvent{
+			Events: unix.EPOLLHUP,
+			Fd:     int32(fd),
+		})
+		if err != nil {
+			log.G(ctx).WithError(err).Warn("notifyClosed: failed to register fd for close notifications")
+			return
+		}
+
+		events := make([]unix.EpollEvent, 1)
+		if _, err := unix_noeintr.EpollWait(epFd, events, -1); err != nil {
```
I don't think `EPOLLHUP` is emitted on a graceful close? Not 100% sure though, but in this case it would cause this goroutine to block indefinitely?

Should this also poll for `EPOLLRDHUP`?

cpuguy83 replied:
If anything using EPOLL_CTL_ADD with EPOLLHUP is not needed because it always looks for EPOLLHUP events.
EPOLLRDHUP is sent when the peer closed the write side.
EPOLLHUP, as I recall, should be when the connection is gone.

vvoland replied:
But I must admit - I don't know exactly how `EPOLLHUP` works, so I'll just trust you if you say that it works as expected 👍🏻

cpuguy83 replied:
I think EPOLLRDHUP would be like the tcp-half close situation which is why a simpler implementation that doesn't require epoll wouldn't work as was discussed above.

thaJeztah replied:
@vvoland this one LGTY?

vvoland replied:
My concern was that `EPOLLHUP` will not be actually emitted on a graceful close based on the C test code.
Just tested the PR though and it looks like it works fine, so 👍🏻

cpuguy83 commented on integration/container/attach_test.go:
```diff
@@ -59,3 +64,58 @@ func TestAttach(t *testing.T) {
+// Regression test for #37182
+func TestAttachDisconnectLeak(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType != "linux", "Bug still exists on Windows")
+	t.Parallel()
```
I had to use a whole separate daemon to get a reliable count.
Since I'm using a separate daemon I made it parallel.

vvoland replied:
Makes sense to me👍🏻

Yinette commented:
Still seeing this in 24.0.3, assuming it's the same issue? Had a system fill itself silly with threads and crashed once it hit kernel limits. I'm 99% sure it's the same issue, however I could be wrong.

```
Package: docker-ce
Version: 5:27.0.3-1~ubuntu.20.04~focal
APT-Sources: https://download.docker.com/linux/ubuntu focal/stable amd64 Packages
```

pprof output:
```
(pprof) text
Showing nodes accounting for 10719, 100% of 10721 total
Dropped 130 nodes (cum <= 53)
Showing top 10 nodes out of 24
      flat  flat%   sum%        cum   cum%
      5487 51.18% 51.18%       5487 51.18%  runtime.gopark
      5232 48.80%   100%       5232 48.80%  syscall.Syscall6
         0     0%   100%       5220 48.69%  github.com/docker/docker/api/server/router/container.notifyClosed
         0     0%   100%       5220 48.69%  github.com/docker/docker/api/server/router/container.notifyClosed.func1
         0     0%   100%       5220 48.69%  github.com/docker/docker/daemon.(*Daemon).containerAttach.func2
         0     0%   100%       5220 48.69%  github.com/docker/docker/internal/unix_noeintr.EpollWait
         0     0%   100%         76  0.71%  github.com/docker/docker/pkg/ioutils.(*BytesPipe).Read
         0     0%   100%         76  0.71%  github.com/docker/docker/pkg/pools.Copy
         0     0%   100%       5220 48.69%  golang.org/x/sys/unix.EpollWait
         0     0%   100%       5220 48.69%  internal/poll.(*FD).RawControl
```

Let me know if you want anything else/or if I've missed something.

Thank you!
